src/CPUActivityAgent.cpp

```diff
@@ -0,0 +1,532 @@
+        if (std::isnan(in_policy[M_POLICY_CPU_PHI])) {
+            in_policy[M_POLICY_CPU_PHI] = M_POLICY_PHI_DEFAULT;
+        }
+
+        if (in_policy[M_POLICY_CPU_PHI] < 0.0 ||
+            in_policy[M_POLICY_CPU_PHI] > 1.0) {
+            throw Exception("CPUActivityAgent::" + std::string(__func__) +
+                                   "(): POLICY_CPU_PHI value out of range: " +
+                                   std::to_string(in_policy[M_POLICY_CPU_PHI]) + ".",
+                                   GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        double f_core_max = in_policy[M_POLICY_CPU_FREQ_MAX];
+        double f_core_efficient = in_policy[M_POLICY_CPU_FREQ_EFFICIENT];
+        double f_core_range = f_core_max - f_core_efficient;
+
+        double f_uncore_max = in_policy[M_POLICY_UNCORE_FREQ_MAX];
+        double f_uncore_efficient = in_policy[M_POLICY_UNCORE_FREQ_EFFICIENT];
+        double f_uncore_range = f_uncore_max - f_uncore_efficient;
+
+        double phi = in_policy[M_POLICY_CPU_PHI];
+
+        // If phi is not 0.5 we move into the energy or performance biased behavior
+        if (phi > 0.5) {
+            // Energy Biased.  Scale F_max down to F_efficient based upon phi value
+            // Active region phi usage
+            f_core_max = std::max(f_core_efficient, f_core_max -
+                                                    f_core_range * (phi-0.5) / 0.5);
+            f_uncore_max = std::max(f_uncore_efficient, f_uncore_max -
+                                                        f_uncore_range * (phi-0.5) / 0.5);
+        }
+        else if (phi < 0.5) {
+            // Perf Biased.  Scale F_efficient up to F_max based upon phi value
+            // Active region phi usage
+            f_core_efficient = std::min(f_core_max, f_core_efficient +
+                                                    f_core_range * (0.5-phi) / 0.5);
+
+            f_uncore_efficient = std::min(f_uncore_max, f_uncore_efficient +
+                                                        f_uncore_range * (0.5-phi) / 0.5);
+        }
+        //Update Policy
+        in_policy[M_POLICY_CPU_FREQ_MAX] = f_core_max;
+        in_policy[M_POLICY_CPU_FREQ_EFFICIENT] = f_core_efficient;
+        in_policy[M_POLICY_UNCORE_FREQ_MAX] = f_uncore_max;
+        in_policy[M_POLICY_UNCORE_FREQ_EFFICIENT] = f_uncore_efficient;
+
+        // Validate all (uncore frequency, max memory bandwidth) pairs
+        std::set<double> policy_uncore_freqs;
+        for (auto it = in_policy.begin() + M_POLICY_FIRST_UNCORE_FREQ;
+             it != in_policy.end() && std::next(it) != in_policy.end(); std::advance(it, 2)) {
+            auto mapped_mem_bw = *(it + 1);
+            if (!std::isnan(*it)) {
+                // We are using a static cast rather than reinterpreting the
+                // memory so that regions can be input to this policy in the
+                // same form they are output from a report.
```

Comment:
An example policy somewhere in the documentation would help explain this comment.  A one line example for the uncore freq policy field right here would also help.